Add manage_containers.py

Willenst opened the pull request:
Hello, it looks like my previous PR #22 was broken, so here is its recovery

a13xp0p0v commented:
Hello @Willenst,

Thanks for your work!

I've added some fixes (including one for #28) and some style improvements.

Currently, I see a possible problem:
I think `Container.check()` should make sure that there is a single container with this pair of compilers.
That would make the code less error-prone.
What do you think about that?

a13xp0p0v commented:
Hi @Willenst,

Could you please also reorder clang and gcc in the `--list` output?

I've added some fixes in `make_linux.py` and renamed it to `build_linux.py`.
It will allow better command autocompletion.

Now you can remove the obsolete shell scripts (because we have `manage_containers.py`). 

And please update the whole README.

Thanks for the collaboration! 👍

Willenst commented:
Hello @a13xp0p0v.

I decided to reorder clang and gcc pretty much everywhere to achieve uniformity, I hope I didn't overdo it. 

Also removed “obsolete” shell scripts and updated README for them and for changes in python scripts.

I also added a couple of beauty features to the README, like text and subparagraph highlights, I hope it will look better this way.

Willenst commented:
Hello @a13xp0p0v!

I fixed the bug with running containers in `Container.rm()`, found a way to pull them directly from the docker without any regulars or greps.

I also demolished Ubuntu 23, it was recently permanently removed from the repositories. 

While poking around in the docker, I came to the conclusion that I need to redo all the naming, because the code actually builds images, not containers, the container is already a directly isolated environment with an image, so I made a global change of names. I ask you to evaluate my changes, after that I can update `README.md` to the current state of the code.

a13xp0p0v commented on manage_containers.py:
```diff
@@ -71,16 +71,17 @@ def add(self):
-        if not running:
+        out = ''
+        cmd = self.runtime_cmd + ['ps', '-a', '--filter', f'ancestor={self.id}', '--format', '{{.ID}}']
+        container_id = subprocess.run(cmd, text=True, check=True, stdout=subprocess.PIPE).stdout.strip()
```
There is a race condition anyway: the container may start just after this check.

How about trying to remove all images without crashing if the container is running?
In this case we can print a message and just proceed to the next image.

a13xp0p0v commented:
Hello @Willenst,

Please see my commits with the improvements.
BTW, I've removed a lot of code :)

Does it work for you?

Willenst commented on manage_containers.py:
```diff
@@ -104,27 +104,27 @@ def identify_runtime_cmd(self):
     for c in images:
         if needed_compiler == 'all':
-            print(f'Adding Ubuntu-{c.ubuntu} container image with Clang-{c.clang} and GCC-{c.gcc}')
+            print(f'Add Ubuntu-{c.ubuntu} container image with Clang-{c.clang} and GCC-{c.gcc}')
```
I think we should also rename Add to Build here and in other places, otherwise this blows the overall logic.

Willenst commented on manage_containers.py:
```diff
@@ -104,27 +104,27 @@ def identify_runtime_cmd(self):
 
 def build_images(needed_compiler, images):
-    """Build container image(s) with the specified compiler"""
+    """Build container image(s) providing the specified compiler"""
```
I think we should edit it to `compilers`, there are a few of them in the back.

Willenst commented on manage_containers.py:
```diff
@@ -106,18 +110,10 @@ def identify_runtime_cmd(self):
             c.build()
-            return
+            if needed_compiler != 'all':
+                return
```
Nice addition, a separate error output for the case with multiple and single addition was really unnecessary! I also liked the technique of checking input below, it was not so obvious to me)

Willenst commented on manage_containers.py:
```diff
@@ -74,12 +74,15 @@ def build(self):
+            subprocess.run(cmd, text=True, check=True)
         except:
-            print("[!] Warning, removal failed! Is image being used?")
+            print("[!] WARNING: image removing failed, see the error message above")
```
By the way, what do you think about a slight visual improvement? I could paint the message in red color, then it will be better to catch the eye in the general pile of logs.

Willenst reviewed (commented):
It's a great tricks to reduce the code, it's a pity that all the beauty of the output has suffered, but from the point of view of architecture, it's a forced sacrifice. Nevertheless, maybe we should add a warning somewhere that user should look at the log above? If the user is not so advanced, he may not realize what is going wrong, because the logs may be quite far away.

Willenst commented:
Hi, @a13xp0p0v !

Thanks for your cool improvements, unfortunately they messed up the tests a bit, but I fixed them. I also found that it is possible to improve the coverage of the tests a bit - implemented this fix. 

Next, I've done a bit of work on the textarea in the code itself, I've also added a small error message output (in case of trying to delete a working container, I hope this will be useful for novice users). 

I also updated the README, I think the moment has come and I decided to revise it now.

Next, I've done a bit of work on the textarea in the code itself, I've also added a small error message output (in case of trying to delete a working container, I hope this will be useful for novice users). 

I also updated the README, I think the moment has come and I decided to revise it now.

a13xp0p0v commented:
Hi @Willenst,
I've finished this pull request and merged the branch (you can see a bunch of commits on top). 
Thanks for the collaboration.
Happy New Year!
